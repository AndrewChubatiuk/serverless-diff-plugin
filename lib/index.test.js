const AWS = require('aws-sdk-mock');
const fs = require('fs-promise');
const Plugin = require('./index');

const slsDir = '.serverless';
const templatePrefix = `${slsDir}/cloudformation-template-update-stack`;
const exampleTemplate = `${templatePrefix}.json`;
const exampleOrgTemplate = `${templatePrefix}.org.json`;
const slsDefaults = {
  service: {
    service: 'foo',
    defaults: {
      stage: 'foo',
      region: 'eu-west-1',
    },
  },
  cli: {
    log: jest.fn(),
  },
};

beforeAll(() => {
  if (!fs.existsSync(slsDir)) {
    fs.mkdirSync(slsDir);
  }
});

describe('serverless-plugin-write-env-vars', () => {
  describe('constructor', () => {
    it('sets the correct defaults', () => {
      const plugin = new Plugin(slsDefaults, {});

      expect(plugin.options.stage).toBe('foo');
      expect(plugin.options.region).toBe('eu-west-1');
      expect(plugin.options.diffTool).toBe('diff');
      expect(plugin.options.localTemplate).toBe(`${templatePrefix}.json`);
      expect(plugin.options.orgTemplate).toBe(`${templatePrefix}.org.json`);
    });

    it('registers the appropriate hooks', () => {
      const plugin = new Plugin(slsDefaults, {});

      expect(typeof plugin.hooks['before:diff:diff']).toBe('function');
      expect(typeof plugin.hooks['diff:diff']).toBe('function');
    });
  });

  describe('downloadTemplate', () => {
    afterEach(() => AWS.restore('CloudFormation'));

    describe('with successful CloudFormation call', () => {
      beforeEach(() =>
        AWS.mock('CloudFormation', 'getTemplate', {
          TemplateBody: '{"foo":"bar"}',
        })
      );

      it('downloads currently deployed template', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.downloadTemplate()
          .then(() =>
            fs.readFile(`${templatePrefix}.org.json`, { encoding: 'utf8' })
              .then(data => expect(data).toMatchSnapshot())
          );
      });
    });

    describe('with unsuccessful CloudFormation call', () => {
      beforeEach(() =>
        AWS.mock('CloudFormation', 'getTemplate', (Object, callback) =>
          callback(new Error('Stack with id foo-foo does not exist'), null)
        )
      );

      it('could not download deployed template', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.downloadTemplate()
          .catch(err => expect(err).toMatchSnapshot());
      });
    });
  });

  describe('diff', () => {
    beforeEach(() =>
      fs.writeFile(exampleTemplate, '{"foo":"foo"}')
        .then(() => fs.writeFile(exampleOrgTemplate, '{"foo":"bar"}'))
    );

    describe('successfully triggers diff', () => {
      it('runs diff with defaults', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.diff()
          .then(data => expect(data).toMatchSnapshot());
      });

      it('runs diff with custom diff tool', () => {
        const plugin = new Plugin(slsDefaults, {
          diffTool: 'diff',
        });

        return plugin.diff()
          .then(data => expect(data).toMatchSnapshot());
      });
    });

    describe('successfully triggers diff', () => {
      const customTemplate = `${templatePrefix}-foo.json`;
      const customOrgTemplate = `${templatePrefix}-foo.org.json`;

      beforeEach(() =>
        fs.writeFile(customTemplate, '{"foo":"foo"}')
          .then(() => fs.writeFile(customOrgTemplate, '{"foo":"custom"}'))
      );

      it('runs diff with custom localTemplate template', () => {
        const plugin = new Plugin(slsDefaults, {
          localTemplate: customTemplate,
        });

        return plugin.diff()
          .then(data => expect(data).toMatchSnapshot());
      });
    });

    describe('unsuccessfully triggers diff', () => {
      beforeEach(() => fs.unlink(exampleTemplate));

      it('could not find locally compiled template', () => {
        const plugin = new Plugin(slsDefaults, {});

        return plugin.diff()
          .catch(err => expect(err).toMatchSnapshot());
      });
    });
  });
});
