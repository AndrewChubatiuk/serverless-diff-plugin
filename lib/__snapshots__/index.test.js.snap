exports[`serverless-plugin-write-env-vars diff successfully triggers diff runs diff with custom diff tool 1`] = `
"1c1
< {\"foo\":\"bar\"}
\\ No newline at end of file
---
> {\"foo\":\"foo\"}
\\ No newline at end of file
"
`;

exports[`serverless-plugin-write-env-vars diff successfully triggers diff runs diff with custom localTemplate template 1`] = `
"1c1
< {\"foo\":\"custom\"}
\\ No newline at end of file
---
> {\"foo\":\"foo\"}
\\ No newline at end of file
"
`;

exports[`serverless-plugin-write-env-vars diff successfully triggers diff runs diff with defaults 1`] = `
"1c1
< {\"foo\":\"bar\"}
\\ No newline at end of file
---
> {\"foo\":\"foo\"}
\\ No newline at end of file
"
`;

exports[`serverless-plugin-write-env-vars diff unsuccessfully triggers diff could not find locally compiled template 1`] = `".serverless/cloudformation-template-update-stack.json could not be found: run \"sls deploy --noDeploy\" first."`;

exports[`serverless-plugin-write-env-vars downloadTemplate with successful CloudFormation call downloads currently deployed template 1`] = `
"{
  \"foo\": \"bar\"
}"
`;

exports[`serverless-plugin-write-env-vars downloadTemplate with unsuccessful CloudFormation call could not download deployed template 1`] = `"Stack with id foo-foo does not exist"`;
